// @tsparticles/plugin-oklch-color/esm/OklchColorManager.js
import { getRangeValue, parseAlpha, } from "@tsparticles/engine";
import { oklchToRgb, oklchaToRgba } from "./utils.js";
export class OklchColorManager {
    constructor() {
        this.key = "color";
        this.stringPrefix = "oklch";
    }
    handleColor(color) {
        const colorValue = color.value, oklchColor = colorValue.oklch ?? color.value;
        if (oklchColor.l !== undefined && oklchColor.c !== undefined && oklchColor.h !== undefined) {
            return oklchToRgb(oklchColor);
        }
    }
    handleRangeColor(color) {
        const colorValue = color.value, oklchColor = colorValue.oklch ?? color.value;
        if (oklchColor.l !== undefined && oklchColor.c !== undefined && oklchColor.h !== undefined) {
            return oklchToRgb({
                l: getRangeValue(oklchColor.l),
                c: getRangeValue(oklchColor.c),
                h: getRangeValue(oklchColor.h),
            });
        }
    }
    parseString(input) {
        const isOklch = input.startsWith("oklch");
        if (!isOklch) {
            return;
        }
        const regex = /oklch\(\s*(\d+(\.\d+)?)%\s+(\d+(\.\d+)?)\s+(\d+(\.\d+)?)(°)?(?:\s*\/\s*(0|1|0?\.\d+|\d{1,3}%))?\s*\)/i, result = regex.exec(input), indexes = {
            l: 1,
            c: 3,
            h: 5,
            a: 7,
        }, defaultAlpha = 1;
        return result
            ? oklchaToRgba({
                a: result[indexes.a] ? parseAlpha(result[indexes.a]) : defaultAlpha,
                c: parseFloat(result[indexes.c]),
                h: parseFloat(result[indexes.h]),
                l: parseFloat(result[indexes.l]),
            })
            : undefined;
    }
}
