// @tsparticles/plugin-oklch-color/esm/utils.js
import { percentDenominator, } from "@tsparticles/engine";
const rgbFactor = 255, fullDegree = 360;
export function lchToRgb(lch) {
    const l = lch.l / percentDenominator, c = lch.c, h = lch.h / fullDegree, result = { r: 0, g: 0, b: 0 };
    result.r = Math.floor(l * rgbFactor);
    result.g = Math.floor(c * rgbFactor);
    result.b = Math.floor(h * rgbFactor);
    return result;
}
export function lchaToRgba(lcha) {
    return {
        a: lcha.a,
        ...lchToRgb(lcha),
    };
}
export function oklchToRgb(oklch) {
    const l = oklch.l / percentDenominator, c = oklch.c / percentDenominator, h = oklch.h / fullDegree, result = { r: 0, g: 0, b: 0 };
    result.r = Math.floor(l * rgbFactor);
    result.g = Math.floor(c * rgbFactor);
    result.b = Math.floor(h * rgbFactor);
    return result;
}
export function oklchaToRgba(oklcha) {
    return {
        a: oklcha.a,
        ...oklchToRgb(oklcha),
    };
}
export function getStyleFromOklch(color, opacity) {
    const { l, c, h } = color, alpha = opacity !== undefined ? `, ${opacity}` : "";
    return `oklch(${l}%, ${c}%, ${h}°${alpha})`;
}
